### PR TITLE
Add final specifier

### LINKED ISSUE #2030: Classes and structs: add final
I think we should. 

_Originally posted by @antkmsft in https://github.com/Azure/azure-sdk-for-cpp/pull/1988#discussion_r605995855_

### PR DESCRIPTION
**Pull request by antkmsft**
Closes #2030.

First, I went and added `final` to everything (excluding Storage[can be a separate PR for that], Nlohmann JSON, non-uinit tests), then I removed everything that did not compile.

I know we discussed to add only the `final` qualifier, but I added a `virtual ~ClassName() = default` in the classes and structs that do not have it, and are inherited. It would be irresponsible to see it and walk past. There are just a couple of instances, and I don't see it creating any problem (other than fixing a memory leak).
To better understand the memory leak problem, see this comment: https://github.com/Azure/azure-sdk-for-cpp/pull/2238#discussion_r628743792

In two instances, `final` specifier is dropped for tests. That preprocessor variable has to be defined when building the library source, not only when building the tests, or otherwise it is not guaranteed, whether the overridden methods defined in tests will get called. Practiacally this means, that when you consume via VcPkg, all the finals are exactly where we want them to be. When you are developing the SDK, two of them are inactive, so you would be able to write a library code that inherits from a final class. But the first nightly vcpkg will fail, and it will be discovered.

### CODE REVIEW

**Comment by antkmsft** [comment]
@vhvb1989 , rather not, it will be technically a breaking change, but removing would not be. Let's make sure now, we have all the data, nothing more is going to change with time.

**Comment by vhvb1989** [comment]
@antkmsft  I got worried based on Casey's recommendation: https://github.com/Azure/azure-sdk-for-cpp/issues/2030#issuecomment-825003808

It looks like it might not be adding too much value but something `micro`.  But, I would leave it to @RickWinter and @JeffreyRichter  to decide if they want to have this in.

**Comment by antkmsft** [comment]
@vhvb1989, this item is scheduled, so we want this, and on Friday sync I was once again assigned this and asked to take care of it.
Plus, this PR fixes a few potential memory leaks with introducing virtual destructor in derived classes.

**Review by vhvb1989 — APPROVED** [review]

File: sdk/core/azure-core/inc/azure/core/base64.hpp [hunk 1]
```diff
@@ -18,7 +18,7 @@ namespace Azure { namespace Core {
    *@brief Used to convert one form of data  into another, for example encoding binary data into
    *base 64 text.
    */
-  class Convert {
+  class Convert final {
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/base64.hpp:21]: All these changes are breaking changes and need a changelog entry. @antkmsft please make sure to add one.

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 2]
```diff
@@ -53,7 +53,7 @@ namespace Azure { namespace Core { namespace Http {
   /**
    * @brief HTTP transport layer error.
    */
-  class TransportException : public Azure::Core::RequestFailedException {
+  class TransportException final : public Azure::Core::RequestFailedException {
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:56]: I don't think we should put final on exception types like these. User's should be able to derive from these to build up the exception heirarchy.

File: sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp [hunk 3]
```diff
@@ -77,7 +77,7 @@ namespace Azure { namespace Core { namespace Credentials {
   /**
    * @brief An exception that gets thrown when authentication error occurs.
    */
-  class AuthenticationException : public std::exception {
+  class AuthenticationException final : public std::exception {
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp:80]: Maybe we should be making an exception for exceptions.

File: sdk/core/azure-core/inc/azure/core/internal/client_options.hpp [hunk 4]
```diff
@@ -24,6 +24,7 @@ namespace Azure { namespace Core { namespace _internal {
    */
   struct ClientOptions
   {
+    virtual ~ClientOptions() = default;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/internal/client_options.hpp:27]: Missing doc comments.

File: sdk/core/azure-core/inc/azure/core/match_conditions.hpp [hunk 5]
```diff
@@ -19,6 +19,8 @@ namespace Azure {
  */
 struct MatchConditions
 {
+  virtual ~MatchConditions() = default;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/match_conditions.hpp:22]: Same here/elsewhere - add doc comments.

Also, please avoid making such changes in a PR that is not related (do it separately). The intent/single purpose of this PR was `Add final specifier`.

Can we add a unit test for this to justify why we needed these virtual dtors?

File: sdk/core/azure-core/src/private/package_version.hpp [hunk 6]
```diff
@@ -20,7 +20,7 @@ namespace Azure { namespace Core { namespace _detail {
   /**
    * @brief Provides version information.
    */
-  class PackageVersion {
+  class PackageVersion final {
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/private/package_version.hpp:23]: Can you share what regex you used to find where to add the `final` keyword?

File: sdk/identity/azure-identity/CMakeLists.txt [hunk 7]
```diff
@@ -26,11 +26,6 @@ if(NOT AZ_ALL_LIBRARIES)
   endif()
 endif()
 
-if(BUILD_TESTING)
-  # define a symbol that enables some test hooks in code
-  add_compile_definitions(TESTING_BUILD)
-endif()
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/CMakeLists.txt:32]: Why are we removing this?

File: sdk/core/azure-core/inc/azure/core/http/http.hpp (hunk 2 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:56]: I don't think we do at this moment view our exception classes being consumed by customers deriving from them.. Our exception is "an exception from Azure SDK", and user's code is not Azure SDK, so I don't see why they should inherit.

But even if this the case, we should be removing final as response to a demand. It is hard to guess what would be needed and what would not, this is why with this change we seal all the ones we are not sure about, leaving only the ones that are intended for inheritance (i.e. base policy), and that are being inherited by our SDK classes.

File: sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp (hunk 3 above)
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/credentials/credentials.hpp:80]: We are starting to think of exception without seeing the need. I, on my side, do not think that customer code should be inheriting from our exceptions - customer code should rather be wrapping them.

File: sdk/core/azure-core/inc/azure/core/io/body_stream.hpp [hunk 8]
```diff
@@ -114,7 +114,7 @@ namespace Azure { namespace Core { namespace IO {
   /**
    * @brief #Azure::Core::IO::BodyStream providing data from an initialized memory buffer.
    */
-  class MemoryBodyStream : public BodyStream {
+  class MemoryBodyStream final : public BodyStream {
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/io/body_stream.hpp:117]: Again, the intent is to remove final where we get requests for it. We do not currently see why a customer would even derive from MemoryBodyStream.

File: sdk/core/azure-core/src/private/package_version.hpp (hunk 6 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/private/package_version.hpp:23]: All done by hand, searching for "`class `" and "`struct `" (with space after).
First I put final everywhere. Then I started removing it until it compiles, while reviewing whether a base class destructor, if public, is virtual.

File: sdk/core/azure-core/test/ut/client_options.cpp [hunk 9]
```diff
@@ -16,7 +16,7 @@ using namespace Azure::Core::_internal;
 using namespace Azure::Core::Http;
 using namespace Azure::Core::Http::Policies;
 
-struct FakeTransport : public HttpTransport
+struct FakeTransport final : public HttpTransport
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/core/azure-core/test/ut/client_options.cpp:19]: That's fine with me.
I had no specific need just consistency and using the same practices in SDK and test code.

File: sdk/identity/azure-identity/CMakeLists.txt (hunk 7 above)
Comment (antkmsft) [reply sdk/identity/azure-identity/CMakeLists.txt:32]: No code depends on `TESTING_BUILD` macro in identity.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp [hunk 10]
```diff
@@ -51,12 +51,18 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
    * Vault. The client supports creating, retrieving, updating, deleting, purging, backing up,
    * restoring, and listing the KeyVaultKey.
    */
-  class KeyClient {
+  class KeyClient
+#if !defined(TESTING_BUILD)
+      final
+#endif
```
[the remark this answers is not in the document]
Comment (antkmsft) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp:57]: mocked_transport_adapter_test.hpp
```
  // A derived class with no credential and authentication
  class KeyClientWithNoAuthenticationPolicy : public Azure::Security::KeyVault::Keys::KeyClient {
  public:
    explicit KeyClientWithNoAuthenticationPolicy(
        std::string const& vaultUrl,
        KeyClientOptions const& options = KeyClientOptions())
        : KeyClient(vaultUrl, nullptr, options)
    {
      auto apiVersion = options.Version.ToString();

      m_pipeline = std::make_unique<Azure::Security::KeyVault::_internal::KeyVaultPipeline>(
          Azure::Core::Url(vaultUrl),
          apiVersion,
          Azure::Core::Http::_internal::HttpPipeline(options, "test", "version", {}, {}));
    }
  };
```

telemetry_header_test.cpp
```
TEST_F(MockedTransportAdapterTest, keyvaultTelemetryId)
{
  std::string applicationId("ourApplicationId");
  m_clientOptions.Telemetry.ApplicationId = applicationId;
  m_client = std::make_unique<
      Azure::Security::KeyVault::Keys::Test::KeyClientWithNoAuthenticationPolicy>(
      "url", m_clientOptions);

  // The fake response from the mocked transport adapter is good for parsing a Key back
  auto response = m_client->GetKey("name");

  // The response is an echo of the sent headers. Let's find the telemetry id
  auto foundHeader = false;
  for (auto& header : response.RawResponse->GetHeaders())
  {
    if (Azure::Core::_internal::StringExtensions::LocaleInvariantCaseInsensitiveEqual(
            header.first, "User-Agent"))
    {
      foundHeader = true;
      EXPECT_PRED2(
          [](std::string const& received, std::string const& sent) {
            auto telemetryInfoWithNoOSAndDate = received.substr(0, sent.size());
            return Azure::Core::_internal::StringExtensions::LocaleInvariantCaseInsensitiveEqual(
                telemetryInfoWithNoOSAndDate, sent);
          },
          header.second,
          applicationId);
      break;
    }
  }
  EXPECT_TRUE(foundHeader);
}
```

This is the only place.

[the remark this answers is not in the document]
Comment (ahsonkhan) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/key_client.hpp:57]: > This is the only test that depended on it. I did not judge whether the test could get away without inheriting

Can you please file an issue for this. We may want to either
1) update the test to avoid the inheritance.
2) use this as a reason to leave keyclient as non-fnal.

cc @vhvb1989
